[ADD] add module to manage product by business

laetitia-gangloff opened the pull request:
First part : add new model (business.product.location) to group some product/quantity into a business and attach these business to a location.

Second part : a report with product/quantity not in the location but excepted and product/quantity not excepted in the location.

Commit:
[ADD] add module to manage product by business

dufresnedavid commented on business_product_location/README.rst:
```diff
@@ -0,0 +1,65 @@
+To begin define a list of product that must be in a stock location.
+Depending on the business usage defined on the location, the list of product
+is different, Locations with the same business usage always have the same
+product list. He also need to compare the theoretical list  with the real
```
who is 'he' ?

dufresnedavid commented on business_product_location/models/business_product_location.py:
```diff
@@ -0,0 +1,160 @@
+                # get product in this location
+                product_in_loc = product_obj.with_context(location=location.id
+                                                          ).search(
+                    [('qty_available', '>', 0)])
```
This syntax is very strange. Please use 2 lines.
product_in_loc = product_obj.with_context(
    location=location.id).search([('qty_available', '>', 0)])

dufresnedavid commented on business_product_location/models/business_product_location.py:
```diff
@@ -0,0 +1,160 @@
+                location_name = '%s - %s' % (location.name,
+                                             location.business_usage_id.name)
+                more_product_by_loc[location_name] = {}
+                less_product_by_loc[location_name] = {}
```
Using a string as dictionary key is most time a bad idea. Nothing makes you sure that you won't somehow have 2 identical tuples of (location.name, location.business_usage_id.name). Also, using a tuple of ids (location.id, location.business_usage_id.id) would be more efficient.

dufresnedavid commented on business_product_location/models/business_product_location.py:
```diff
@@ -0,0 +1,160 @@
+                            product.id] = product_to_update
+                    else:
+                        less_product_by_loc[location_name][
+                            product.id] = product_to_update
```
Consider split your method (render_html) into different functions. The indentation makes it hard to read.

Commit:
[ADD] business_product_location : add report

Add a report to get the difference between theorical product and real product at the location

laetitia-gangloff commented:
@dufresnedavid thank you for your suggestions !

I splitted render_html and tried to add more comments. For the key, I thought location name was unique, but it is not the case, so I updated.

ghost commented:
@laetitia-gangloff Thanks :+1:

lmignon commented on business_product_location/__openerp__.py:
```diff
@@ -0,0 +1,37 @@
+{
+    "name": "Business Product Location",
+    "version": "0.1",
+    'author': "Acsone, Odoo Community Association (OCA)",
```
ACSONE SA/NV

lmignon commented on business_product_location/__openerp__.py:
```diff
@@ -0,0 +1,37 @@
+    "name": "Business Product Location",
+    "version": "0.1",
+    'author': "Acsone, Odoo Community Association (OCA)",
+    "category": "Other",
```
'category': 'Warehouse Management',

lmignon commented on business_product_location/models/business_product_location.py:
```diff
@@ -0,0 +1,192 @@
+                                        'Business Usage')
+
+
+class BusinessProductLocationReport(models.AbstractModel):
```
According to the OCA guidelines, reports should be in a report module. https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#directories

lmignon commented on business_product_location/tests/test_business_product_location.py:
```diff
@@ -0,0 +1,68 @@
+
+class TestBusinessProductLocation(common.TransactionCase):
+
+    def setUp(self):
```
This override is not needed.

lmignon commented on business_product_location/tests/test_business_product_location.py:
```diff
@@ -0,0 +1,68 @@
+        bpl = self.env['business.product.location'].create(
+            {'name': 'Test BPL',
+             'product_ids': [
+                 (0, 0, {'product_id': self.env.ref(
```
PLZ assign self.env.ref('product.product_product_3X) to a variable to avoir copy/paste and improve the readability

lmignon commented on business_product_location/tests/test_business_product_location.py:
```diff
@@ -0,0 +1,68 @@
+                 (0, 0, {'product_id': self.env.ref(
+                     'product.product_product_35').id, 'product_qty': 3})],
+             'location_ids': [(4,
+                               self.env.ref('stock.stock_location_stock').id)]
```
Idem

lmignon commented on business_product_location/views/business_product_location_templates.xml:
```diff
@@ -0,0 +1,69 @@
+                    <t t-call="report.external_layout">
+                        <div class="page">
+                            <h2>
+                            Excepted product - <span t-esc="o_value['location_name']"/>
```
Is this label translatable?

laetitia-gangloff replied:
It uses the lang of the current user.

lmignon commented:
The functionality is interesting. A better description would be welcome.

Commit:
[IMP] business_product_location : update test readability / use folder report / update __openerp__.py

gurneyalex commented:
can you please rebase your PR? The travis build of 8.0 has recently been fixed.

laetitia-gangloff commented:
It is rebased. Thank you.

gurneyalex commented:
:+1:

Cedric-Pigeon commented:
+1

Cedric-Pigeon commented:
@pedrobaeza I think this one can be merged

lmignon reviewed (approved):
:+1: Works in production for a long time

pedrobaeza commented:
README file is outdated, but I won't be picky.
